server/services/nextcloud-talk/lib/message/message.bot.js

```diff
@@ -0,0 +1,77 @@
+const EventEmitter = require('events');
+const get = require('get-value');
+const logger = require('../../../../utils/logger');
+
+class NextcloudTalkBot extends EventEmitter {
+  constructor(gladys, serviceId, userId, token) {
+    super();
+    this.gladys = gladys;
+    this.serviceId = serviceId;
+    this.userId = userId;
+    this.token = token;
+    this.isPolling = false;
+  }
+
+  async startPolling() {
+    logger.debug(`Start polling Nextcloud Talk for token: ${this.token}`);
+
+    const user = await this.gladys.user.getById(this.userId);
+
+    if (!this.NEXTCLOUD_URL) {
+      this.NEXTCLOUD_URL = await this.gladys.variable.getValue('NEXTCLOUD_URL', this.serviceId, user.id);
+    }
+    if (!this.NEXTCLOUD_BOT_USERNAME) {
+      this.NEXTCLOUD_BOT_USERNAME = await this.gladys.variable.getValue(
+        'NEXTCLOUD_BOT_USERNAME',
+        this.serviceId,
+        user.id,
+      );
+    }
+    if (!this.NEXTCLOUD_BOT_PASSWORD) {
+      this.NEXTCLOUD_BOT_PASSWORD = await this.gladys.variable.getValue(
+        'NEXTCLOUD_BOT_PASSWORD',
+        this.serviceId,
+        user.id,
+      );
+    }
+
+    this.isPolling = true;
+    this.poll();
+  }
+
+  stopPolling() {
+    logger.debug(`Stop polling Nextcloud Talk for token: ${this.token}`);
+    this.isPolling = false;
+  }
+
+  async poll(lastKnownMessageId) {
+    const result = await this.gladys.http.request(
+      'get',
+      `${this.NEXTCLOUD_URL}/ocs/v2.php/apps/spreed/api/v1/chat/${this.token}?lookIntoFuture=1&timeout=15${
+        lastKnownMessageId ? `&lastKnownMessageId=${lastKnownMessageId}` : ''
+      }`,
+      '',
+      {
+        Authorization: `Basic ${Buffer.from(`${this.NEXTCLOUD_BOT_USERNAME}:${this.NEXTCLOUD_BOT_PASSWORD}`).toString(
+          'base64',
+        )}`,
+        Accept: 'application/json',
+        'OCS-APIRequest': true,
+      },
+    );
+    const newMessages = get(result, 'data.ocs.data');
+    if (
+      lastKnownMessageId &&
+      newMessages &&
+      newMessages[newMessages.length - 1].actorId !== this.NEXTCLOUD_BOT_USERNAME
+    ) {
+      this.emit('message', newMessages.pop());
+    }
+
+    if (this.isPolling) {
+      setTimeout(() => this.poll(result.headers['x-chat-last-given'] || lastKnownMessageId), 50);
```

Comment:
Are you polling every 50ms forever? It's a bit extreme no?

Reply:
It's not really that, it requests for new message here https://github.com/GladysAssistant/Gladys/pull/1388/files#diff-388fb960051d373b1c5591fe8a071c088ff028c5d4628956265ebbbadc6b6b3bR16-R20
There is a timeout define in the request query parameters to 15, the request end on new message response OR on timeout, then retry if we polling is enable, I added a delay of 50ms to possibly reduce load but it's not really necessary.

Reply:
But ` this.bots[userId].isPolling = true` right ?

So it's always doing a request, wait 50ms, one other request?

Considering the request takes 100ms, it means we are requesting every 150ms forever

It means 6.6 req/seconds = 2160 req/min = 129k requests per hour = 3 millions requests per day = 90 millions req/month

It's wayyy to much 😅

Reply:
Ok makes sense, I went reading the code and indeed!